File: backend/penguin_mail/api/schemas/account.py
from datetime import datetime

from ninja import Schema


class AccountOut(Schema):
    id: str
    email: str
    name: str
    color: str
    provider: str
    displayName: str
    signature: str
    defaultSignatureId: str
    avatar: str
    isDefault: bool
    createdAt: datetime
    updatedAt: datetime
    lastSyncAt: datetime | None = None

    @staticmethod
    def from_model(account) -> "AccountOut":
        return AccountOut(
            id=str(account.uuid),
            email=account.email,
            name=account.name,
            color=account.color,
            provider=account.provider,
            displayName=account.display_name,
            signature=account.signature,
            defaultSignatureId=account.default_signature_id,
            avatar=account.avatar,
            isDefault=account.is_default,
            createdAt=account.created_at,
            updatedAt=account.updated_at,
            lastSyncAt=account.last_sync_at,
        )


class AccountCreateIn(Schema):
    email: str
    name: str
    color: str = "blue"
    displayName: str = ""
    signature: str = ""

    # Provider-based setup
    provider: str = "custom"
    password: str = ""

    # Custom server fields (only required when provider is 'custom')
    smtp_host: str | None = None
    smtp_port: int | None = None
    smtp_security: str | None = None
    imap_host: str | None = None
    imap_port: int | None = None
    imap_security: str | None = None


class AccountUpdateIn(Schema):
    name: str | None = None
    color: str | None = None
    displayName: str | None = None
    signature: str | None = None
    defaultSignatureId: str | None = None
    avatar: str | None = None
    isDefault: bool | None = None


class TestConnectionIn(Schema):
    email: str
    provider: str = "custom"
    password: str = ""
    smtp_host: str | None = None
    smtp_port: int | None = None
    smtp_security: str | None = None
    imap_host: str | None = None
    imap_port: int | None = None
    imap_security: str | None = None


class TestConnectionOut(Schema):
    smtp: bool
    imap: bool
    smtp_error: str = ""
    imap_error: str = ""

File: frontend/src/types/account.ts
/**
 * Account types for the email client
 */

import type { SystemFolderType } from './email'
import type { EmailProvider } from '@/constants/providers'

export type AccountColor =
  | 'blue'
  | 'green'
  | 'purple'
  | 'orange'
  | 'pink'
  | 'teal'
  | 'red'
  | 'indigo'

export interface Account {
  id: string
  email: string
  name: string
  color: AccountColor
  provider: string
  displayName?: string
  signature?: string
  defaultSignatureId?: string
  avatar?: string
  isDefault: boolean
  createdAt: Date
  updatedAt: Date
  lastSyncAt?: Date | null
}

export interface AccountCreateInput {
  email: string
  name: string
  color?: AccountColor
  displayName?: string
  signature?: string

  provider: EmailProvider
  password: string

  // Only for custom provider
  smtp_host?: string
  smtp_port?: number
  smtp_security?: string
  imap_host?: string
  imap_port?: number
  imap_security?: string
}

export interface AccountUpdateInput {
  name?: string
  color?: AccountColor
  displayName?: string
  signature?: string
  defaultSignatureId?: string
  avatar?: string
  isDefault?: boolean
}

export const ACCOUNT_COLOR_VAR: Record<AccountColor, string> = {
  blue: 'var(--account-blue)',
  green: 'var(--account-green)',
  purple: 'var(--account-purple)',
  orange: 'var(--account-orange)',
  pink: 'var(--account-pink)',
  teal: 'var(--account-teal)',
  red: 'var(--account-red)',
  indigo: 'var(--account-indigo)',
}

export const FOLDER_LABELS: Record<SystemFolderType, string> = {
  inbox: 'Inbox',
  drafts: 'Drafts',
  sent: 'Sent',
  spam: 'Spam',
  trash: 'Trash',
  archive: 'Archive',
  starred: 'Starred',
  snoozed: 'Snoozed',
  scheduled: 'Scheduled',
}
